Add initial support for the browser.i18n Web Extension API.

kiaraarose opened the pull request:
#### f756fa5b5b9132fec9add4ec28cb6f91c1d70409
<pre>
Add initial support for the browser.i18n Web Extension API.
<a href="https://webkit.org/b/246488">https://webkit.org/b/246488</a>
rdar://102720462 (Web Extensions in WebKit: Add support for i18n(Localization) in WebKit)

Reviewed by Timothy Hatcher.

This patch exposes the i18n API to the browser namespace. In addition, it add support for the methods:
- i18n.getAcceptLanguages()
- i18n.getUILanguage()

To do:
- Add support for i18n.getMessage().

Testing:
- Localized names for extensions show up in UI.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::dictionaryWithLowercaseKeys):
(WebKit::mergeDictionaries): Helper method that merges two dictionaries together.
(WebKit::mergeDictionariesAndSetValues): Helper method that merges two dictionaries together,
overwriting existing values.
(WebKit::localeStringInWebExtensionFormat):
(WebKit::privacyPreservingDescription):
(WebKit::toImpl):
(WebKit::toImplSet):

* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.h: Added.
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm: Added.
(localizationDictionaryAtURL):
(+[_WKWebExtensionLocalization supportsSecureCoding]):
(-[_WKWebExtensionLocalization initWithBundleURL:defaultLocale:uniqueIdentifier:]):
(-[_WKWebExtensionLocalization initWithRegionalLocalization:languageLocalization:defaultLocalization:withBestLocale:uniqueIdentifier:]):
(-[_WKWebExtensionLocalization localizedDictionaryForDictionary:]):
(-[_WKWebExtensionLocalization localizedStringForKey:withPlaceholders:]):
(-[_WKWebExtensionLocalization _localizedArrayForArray:]):
(-[_WKWebExtensionLocalization localizedStringForString:]):
(-[_WKWebExtensionLocalization _predefinedMessagesForLocale:]):
(-[_WKWebExtensionLocalization _stringByReplacingNamedPlaceholdersInString:withNamedPlaceholders:]):
(-[_WKWebExtensionLocalization _stringByReplacingPositionalPlaceholdersInString:withValues:]):
(-[_WKWebExtensionLocalization encodeWithCoder:]):
(-[_WKWebExtensionLocalization initWithCoder:]):

* Source/WebKit/Shared/Extensions/_WKWebExtensionUtilities.mm:
(+[_WKWebExtensionUtilities validateContentsOfDictionary:requiredKeys:optionalKeys:keyToExpectedValueType:outExceptionString:]):
Add error message.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension defaultLocale]):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
Update FIXME with appropriate bug URL.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::manifest):
(WebKit::WebExtension::localization):
(WebKit::WebExtension::defaultLocale):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const):
FIXME to add localization dictionary to parameters struct.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm:

(WebKit::WebExtensionAPIExtension::getURL):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm: Added.
(WebKit::WebExtensionAPILocalization::getMessage):
(WebKit::WebExtensionAPILocalization::getUILanguage):
(WebKit::WebExtensionAPILocalization::getAcceptLanguages):

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::i18n):
Add i18n to the browser namespace.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
Add links to documentation for Permissions APIs.
(WebKit::WebExtensionAPIPermissions::getAll):
(WebKit::WebExtensionAPIPermissions::contains):
(WebKit::WebExtensionAPIPermissions::request):
(WebKit::WebExtensionAPIPermissions::remove):
(WebKit::WebExtensionAPIPermissions::onAdded):
(WebKit::WebExtensionAPIPermissions::onRemoved):

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm:
Add links to documentation for Web Navigation APIs.
(WebKit::WebExtensionAPIWebNavigation::onBeforeNavigate):
(WebKit::WebExtensionAPIWebNavigation::onCommitted):
(WebKit::WebExtensionAPIWebNavigation::onDOMContentLoaded):
(WebKit::WebExtensionAPIWebNavigation::onCompleted):
(WebKit::WebExtensionAPIWebNavigation::onErrorOccurred):
(WebKit::WebExtensionAPIWebNavigation::getAllFrames):
(WebKit::WebExtensionAPIWebNavigation::getFrame):

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPILocalization.h: Copied from Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPILocalization.idl: Copied from Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
Add i18n to the browser namespace.

Canonical link: <a href="https://commits.webkit.org/267316@main">https://commits.webkit.org/267316@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e15f1c1690565abdb998eeda87cc5bdb2a9aac50

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16228 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16546 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16961 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17994 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15228 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19623 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16659 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17621 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16422 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16870 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13872 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18759 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14113 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21523 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15102 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14854 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/18093 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15449 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13099 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14675 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19041 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1997 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15272 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -186,6 +188,120 @@
+    static NSSet<NSString *> *identifiers;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        identifiers = [NSSet setWithArray:@[
```
should check to see if this list needs updating

kiaraarose replied:
actually i'm removing this since it's not needed

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    NSString *returnedMessage;
+    // FIXME: <https://webkit.org/b/246488> Retrieve localized dictionary from the WebExtensionContextProxy.
```
@xeenon is this correct?

xeenon replied:
Yeah

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -186,6 +188,120 @@
+            mergedDictionary[key] = dictionaryB[key];
+    }
+
+    return mergedDictionary;
```
`return [mergedDictionary copy]`

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -186,6 +188,120 @@
+    for (id key in dictionaryB.allKeys)
+        newDictionary[key] = dictionaryB[key];
+
+    return newDictionary;
```
`return [mergedDictionary copy]`

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -186,6 +188,120 @@
+
+// MARK: NSLocale helper methods.
+
+NSSet *localeIdentifiersForMostWidelyUsedLanguages()
```
Where is this used?

kiaraarose replied:
absolutely nowhere. added by accident. will remove

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -253,4 +369,26 @@ WallTime toImpl(NSDate *date)
+    result.reserveInitialCapacity(array.count);
+
+    for (NSString *element in array)
+        result.append(element);
```
`appendUnchecked()` since you use `reserveInitialCapacity`

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -253,4 +369,26 @@ WallTime toImpl(NSDate *date)
+    result.reserveInitialCapacity(array.count);
+
+    for (NSString *element in array)
+        result.add(element);
```
`addVoid()` since you don't use the result of `add()`.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,334 @@
+static NSString * const localeCodingKey = @"locale";
+static NSString * const uniqueIdentifierCodingKey = @"uniqueIdentifier";
+
+typedef NSDictionary<NSString *, NSDictionary *> LocalizationDictionary;
```
`using LocalizationDictionary = NSDictionary<NSString *, NSDictionary *>`

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,334 @@
+static NSString * const uniqueIdentifierCodingKey = @"uniqueIdentifier";
+
+typedef NSDictionary<NSString *, NSDictionary *> LocalizationDictionary;
+typedef NSDictionary<NSString *, NSDictionary<NSString *, NSString *> *> PlaceholderDictionary;
```
`using PlaceholderDictionary = NSDictionary<NSString *, NSDictionary<NSString *, NSString *> *>`

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,334 @@
+NSString * const WKPredefinedMessageValueLeftToRight = @"ltr";
+NSString * const WKPredefinedMessageValueRightToLeft = @"rtl";
+NSString * const WKPredefinedMessageValueTextEdgeLeft = @"left";
+NSString * const WKPredefinedMessageValueTextEdgeRight = @"right";
```
Most of these are only used in this file, they should be `static` then and drop the `WK` prefix.

kiaraarose replied:
I kept the first three as with the WK since I plan on using them for testing, but changed the rest

xeenon replied:
They should be `_WK` prefix then since they are private.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,334 @@
+        return nil;
+
+    NSError *jsonParsingError;
+    LocalizationDictionary *localizationDictionary = (NSDictionary *)[NSJSONSerialization JSONObjectWithData:dictionaryData options:0 error:&jsonParsingError];
```
This should use `dynamic_objc_cast<NSDictionary>()` instead of `(NSDictionary *)`, since it isn't guaranteed to be a dictionary.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,334 @@
+
+    PlaceholderDictionary *namedPlaceholders = dictionaryWithLowercaseKeys(objectForKey<NSDictionary>(stringDictionary, WKLocalizationDictionaryPlaceholdersKey));
+
+
```
Remove extra newline.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,334 @@
+    return self;
+}
+
+
```
Remove extra newline.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -141,11 +143,15 @@
+    ASSERT([NSJSONSerialization isValidJSONObject:manifest]);
+
+    NSData *manifestData = [NSJSONSerialization dataWithJSONObject:manifest options:0 error:nullptr];
+    ASSERT(manifestData);
```
`RELEASE_ASSERT` since this will crash if `nil` when doing `setObject:` below. Might as well crash where the problem is instead.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.h:
```diff
@@ -75,12 +77,29 @@ T *objectForKey(const RetainPtr<NSDictionary>& dictionary, id key, bool returnin
+
+NSDictionary *dictionaryWithLowercaseKeys(NSDictionary *);
+NSDictionary *mergeDictionaries(NSDictionary *, NSDictionary *);
+NSDictionary *mergeAndSetValues(NSDictionary *, NSDictionary *);
```
Either name the first one `merge()` or the second one `mergeDictionariesAndSetValues()`.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -186,6 +188,92 @@
+
+    return [NSError errorWithDomain:error.domain ?: @"" code:error.code userInfo:nil].description;
+}
+
```
I learned something new the other day. `NSErrors` will be redacted automatically by `os_log` when used in a plain `%@`  or `%{private}@` placeholder by calling the `redactedDescription` method on the `NSError` object (and many other objects). We only need to do this if we want to log the error with `%{public}@`.

xeenon replied:
I also don'y see you using this anywhere, so I would just remove it for now. I think logging the error as `%{private}@` is fine for now.

kiaraarose replied:
This is being used on line 82 in `_WKWebExtensionLocalization.mm`

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -32,6 +32,8 @@
 
 namespace WebKit {
 
+NSString * const WebkitPrivacyPreservingDescription = @"WebkitPrivacyPreservingDescription";
```
`static NSString * const privacyPreservingDescriptionKey = @"WebKitPrivacyPreservingDescription";`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
```diff
@@ -0,0 +1,97 @@
+{
+    // Documentation: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/i18n/getMessage
+
+    RELEASE_LOG_INFO(Extensions, "i18n.getMessage()");
```
I posted a [PR](https://github.com/WebKit/WebKit/pull/16854) that added these logs to the code generator, so we don't need to put them in manually anymore.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.h:
```diff
@@ -0,0 +1,52 @@
+
+@property (nonatomic, nullable, copy) NSString *uniqueIdentifier;
+
+- (instancetype)initWithBundleURL:(NSURL *)bundleURL defaultLocale:(NSString *)defaultLocaleString uniqueIdentifier:(nullable NSString *)uniqueIdentifier;
```
Needs to be `NS_DESIGNATED_INITIALIZER` too to fix tvOS wand watchOS builds.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.h:
```diff
@@ -0,0 +1,52 @@
+
+- (instancetype)initWithBundleURL:(NSURL *)bundleURL defaultLocale:(NSString *)defaultLocaleString uniqueIdentifier:(nullable NSString *)uniqueIdentifier;
+- (instancetype)initWithRegionalLocalization:(nullable NSDictionary<NSString *, NSDictionary *> *)regionalLocalization languageLocalization:(nullable NSDictionary<NSString *, NSDictionary *> *)languageLocalization defaultLocalization:(nullable NSDictionary<NSString *, NSDictionary *> *)defaultLocalization withBestLocale:(nullable NSString *)localeString uniqueIdentifier:(NSString *)uniqueIdentifi ...
+- (nullable instancetype)initWithCoder:(NSCoder *)coder NS_DESIGNATED_INITIALIZER;
```
Can be removed, and add `NSSecureCoding` above.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.h:
```diff
@@ -0,0 +1,52 @@
+WK_EXTERN NSString * const _WKLocalizationDictionaryPlaceholdersKey;
+
+WK_EXTERN
+@interface _WKWebExtensionLocalization : NSObject
```
Add `<NSSecureCoding>` to the end of the line here.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,372 @@
++ (BOOL)supportsSecureCoding
+{
+    return YES;
+}
```
Move to the top, outside the `#if ENABLE(WK_WEB_EXTENSIONS)` but just inside the `@implementation`.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,372 @@
+    return YES;
+}
+
+- (void)encodeWithCoder:(nonnull NSCoder *)coder
```
Drop `nonnull`.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,372 @@
+        [coder encodeObject:_uniqueIdentifier forKey:uniqueIdentifierCodingKey];
+}
+
+- (instancetype)initWithCoder:(nonnull NSCoder *)coder
```
Drop `nonnull`.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,372 @@
+{
+    return nil;
+}
+
```
Need to add empty `- (void)encodeWithCoder:(NSCoder *)coder` here.

kiaraarose commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,377 @@
+    if (!!defaultLocaleDictionary)
+        bestLocaleString = defaultLocaleString;
+    else
+        RELEASE_LOG(Extensions, "Could not find localization for %{public}@ for extension bundle at URL %{private}@", defaultLocaleString, bundleURL);
```
`RELEASE_LOG_ERROR`, and the same for those below

xeenon replied:
I think `RELEASE_LOG_DEBUG` are fine here, these are not really errors just attempts to find the right file.

xeenon replied:
In fact I think a single `RELEASE_LOG_DEBUG` after all the if lines would be best, logging what `bestLocaleString` was picked not what couldn't be found.

kiaraarose commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,377 @@
+
+    return self;
+}
+
```
remove newline

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,377 @@
+    NSString *bestLocaleString;
+
+    LocalizationDictionary *defaultLocaleDictionary = localizationDictionaryAtURL(defaultLocaleDictionaryURL);
+    if (!!defaultLocaleDictionary)
```
Nit: Can remove the `!!` on all these lines.

kiaraarose commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,377 @@
+
+- (void)encodeWithCoder:(NSCoder *)coder
+{
+    return nil;
```
don't return nil

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,375 @@
+
+    return self;
+}
+#else
```
Nit: Add newline above.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -0,0 +1,375 @@
+
+- (instancetype)initWithCoder:(NSCoder *)coder
+{
+    return nil;
```
Try just adding: `return [self initWithRegionalLocalization:nil languageLocalization:nil defaultLocalization:nil withBestLocale:nil uniqueIdentifier:@""];` here to fix the build.

kiaraarose committed:
Add initial support for the browser.i18n Web Extension API.
https://webkit.org/b/246488
rdar://102720462 (Web Extensions in WebKit: Add support for i18n(Localization) in WebKit)

Reviewed by Timothy Hatcher.

This patch exposes the i18n API to the browser namespace. In addition, it add support for the methods:
- i18n.getAcceptLanguages()
- i18n.getUILanguage()

To do:
- Add support for i18n.getMessage().

Testing:
- Localized names for extensions show up in UI.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::dictionaryWithLowercaseKeys):
(WebKit::mergeDictionaries): Helper method that merges two dictionaries together.
(WebKit::mergeDictionariesAndSetValues): Helper method that merges two dictionaries together,
overwriting existing values.
(WebKit::localeStringInWebExtensionFormat):
(WebKit::privacyPreservingDescription):
(WebKit::toImpl):
(WebKit::toImplSet):

* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.h: Added.
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm: Added.
(localizationDictionaryAtURL):
(+[_WKWebExtensionLocalization supportsSecureCoding]):
(-[_WKWebExtensionLocalization initWithBundleURL:defaultLocale:uniqueIdentifier:]):
(-[_WKWebExtensionLocalization initWithRegionalLocalization:languageLocalization:defaultLocalization:withBestLocale:uniqueIdentifier:]):
(-[_WKWebExtensionLocalization localizedDictionaryForDictionary:]):
(-[_WKWebExtensionLocalization localizedStringForKey:withPlaceholders:]):
(-[_WKWebExtensionLocalization _localizedArrayForArray:]):
(-[_WKWebExtensionLocalization localizedStringForString:]):
(-[_WKWebExtensionLocalization _predefinedMessagesForLocale:]):
(-[_WKWebExtensionLocalization _stringByReplacingNamedPlaceholdersInString:withNamedPlaceholders:]):
(-[_WKWebExtensionLocalization _stringByReplacingPositionalPlaceholdersInString:withValues:]):
(-[_WKWebExtensionLocalization encodeWithCoder:]):
(-[_WKWebExtensionLocalization initWithCoder:]):

* Source/WebKit/Shared/Extensions/_WKWebExtensionUtilities.mm:
(+[_WKWebExtensionUtilities validateContentsOfDictionary:requiredKeys:optionalKeys:keyToExpectedValueType:outExceptionString:]):
Add error message.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension defaultLocale]):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
Update FIXME with appropriate bug URL.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::manifest):
(WebKit::WebExtension::localization):
(WebKit::WebExtension::defaultLocale):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const):
FIXME to add localization dictionary to parameters struct.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm:

(WebKit::WebExtensionAPIExtension::getURL):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm: Added.
(WebKit::WebExtensionAPILocalization::getMessage):
(WebKit::WebExtensionAPILocalization::getUILanguage):
(WebKit::WebExtensionAPILocalization::getAcceptLanguages):

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::i18n):
Add i18n to the browser namespace.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
Add links to documentation for Permissions APIs.
(WebKit::WebExtensionAPIPermissions::getAll):
(WebKit::WebExtensionAPIPermissions::contains):
(WebKit::WebExtensionAPIPermissions::request):
(WebKit::WebExtensionAPIPermissions::remove):
(WebKit::WebExtensionAPIPermissions::onAdded):
(WebKit::WebExtensionAPIPermissions::onRemoved):

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm:
Add links to documentation for Web Navigation APIs.
(WebKit::WebExtensionAPIWebNavigation::onBeforeNavigate):
(WebKit::WebExtensionAPIWebNavigation::onCommitted):
(WebKit::WebExtensionAPIWebNavigation::onDOMContentLoaded):
(WebKit::WebExtensionAPIWebNavigation::onCompleted):
(WebKit::WebExtensionAPIWebNavigation::onErrorOccurred):
(WebKit::WebExtensionAPIWebNavigation::getAllFrames):
(WebKit::WebExtensionAPIWebNavigation::getFrame):

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPILocalization.h: Copied from Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPILocalization.idl: Copied from Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
Add i18n to the browser namespace.

Canonical link: https://commits.webkit.org/267316@main
